Mathlib/Algebra/Algebra/Subalgebra/MulOpposite.lean

```diff
@@ -0,0 +1,172 @@
+/-
+Copyright (c) 2024 Jz Pan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jz Pan
+-/
+import Mathlib.Algebra.Ring.Subring.MulOpposite
+import Mathlib.Algebra.Algebra.Subalgebra.Basic
+
+/-!
+
+# Subalgebras of opposite rings
+
+For every ring `A` over a commutative ring `R`, we construct an equivalence between
+subalgebras of `A / R` and that of `Aᵐᵒᵖ / R`.
+
+-/
+
+namespace Subalgebra
+
+section Semiring
+
+variable {ι R A : Type*} [CommSemiring R] [Semiring A] [Algebra R A]
+
+/-- Pull a subalgebra back to an opposite subalgebra along `MulOpposite.unop` -/
+@[simps toSubsemiring]
+protected def op (S : Subalgebra R A) : Subalgebra R Aᵐᵒᵖ where
+  toSubsemiring := S.toSubsemiring.op
+  algebraMap_mem' := S.algebraMap_mem
+
+@[simp]
+theorem op_coe (S : Subalgebra R A) : S.op = MulOpposite.unop ⁻¹' (S : Set A) := rfl
+
+@[simp]
+theorem mem_op {x : Aᵐᵒᵖ} {S : Subalgebra R A} : x ∈ S.op ↔ x.unop ∈ S := Iff.rfl
+
+/-- Pull an subalgebra subring back to a subalgebra along `MulOpposite.op` -/
+@[simps toSubsemiring]
+protected def unop (S : Subalgebra R Aᵐᵒᵖ) : Subalgebra R A where
+  toSubsemiring := S.toSubsemiring.unop
+  algebraMap_mem' := S.algebraMap_mem
+
+@[simp]
+theorem unop_coe (S : Subalgebra R Aᵐᵒᵖ) : S.unop = MulOpposite.op ⁻¹' (S : Set Aᵐᵒᵖ) := rfl
+
+@[simp]
+theorem mem_unop {x : A} {S : Subalgebra R Aᵐᵒᵖ} : x ∈ S.unop ↔ MulOpposite.op x ∈ S := Iff.rfl
+
+@[simp]
+theorem unop_op (S : Subalgebra R A) : S.op.unop = S := rfl
+
+@[simp]
+theorem op_unop (S : Subalgebra R Aᵐᵒᵖ) : S.unop.op = S := rfl
+
+/-! ### Lattice results -/
+
+theorem op_le_iff {S₁ : Subalgebra R A} {S₂ : Subalgebra R Aᵐᵒᵖ} : S₁.op ≤ S₂ ↔ S₁ ≤ S₂.unop :=
+  MulOpposite.op_surjective.forall
+
+theorem le_op_iff {S₁ : Subalgebra R Aᵐᵒᵖ} {S₂ : Subalgebra R A} : S₁ ≤ S₂.op ↔ S₁.unop ≤ S₂ :=
+  MulOpposite.op_surjective.forall
+
+@[simp]
+theorem op_le_op_iff {S₁ S₂ : Subalgebra R A} : S₁.op ≤ S₂.op ↔ S₁ ≤ S₂ :=
+  MulOpposite.op_surjective.forall
+
+@[simp]
+theorem unop_le_unop_iff {S₁ S₂ : Subalgebra R Aᵐᵒᵖ} : S₁.unop ≤ S₂.unop ↔ S₁ ≤ S₂ :=
+  MulOpposite.unop_surjective.forall
+
+/-- A subalgebra `S` of `A / R` determines a subring `S.op` of the opposite ring `Aᵐᵒᵖ / R`. -/
+@[simps]
+def opEquiv : Subalgebra R A ≃o Subalgebra R Aᵐᵒᵖ where
```

Comment:
```suggestion
def opOrderIso : Subalgebra R A ≃o Subalgebra R Aᵐᵒᵖ where
```

Reply:
But it's called [Submonoid.opEquiv](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/Submonoid/MulOpposite.html#Submonoid.opEquiv) for Submonoid, etc.

Reply:
Uh, okay 😦